### ISSUE TITLE
ModelObserver#onTransactionStart not called when the operation starts, only when it's committed

### ISSUE THREAD

**Issue by merwaaan** [SketchUp/api-issue-tracker#704]
# Issue

1. SketchUp/LayOut Version: 2017, 2021.1
2. OS Platform: Windows

Hi,
I have another problem with operation-related observers (see #703). 
Opening a new issue in order not to mix things up.

It seems that `onTransactionStart` is not called when starting an operation via `model.start_operation`.
However, it's called when using `model.commit_operation`.

Here's a reproducer:

```ruby
class ModelObserver < Sketchup::ModelObserver

  def onTransactionStart(_model)
    puts 'start'
  end

  def onTransactionCommit(_model)
    puts 'commit'
  end
end

Sketchup.active_model.add_observer(ModelObserver.new)

puts "before start"
Sketchup.active_model.start_operation('test')
puts "after start"

Sketchup.active_model.active_entities.add_group

puts "before commit"
Sketchup.active_model.commit_operation
puts "after commit"
```

I expected to get
```
before start
start
after start
before commit
commit
after commit
```

but instead I got
```
before start
after start
before commit
start
commit
after commit
```

Is that expected behavior?
I expected the onTransactionXXX event handlers to match the XXX_operation methods.

---

# My use case

I'm writing a wrapper to make SketchUp operations a little bit safer, with an API like

```
ModelOperation.do(some options) {
  # update the model here
}
```

Internally it uses the SketchUp API to start/commit an operation, executes the block in between those steps, does some logging, aborts the operation in case of error etc.

I also want this wrapper to watch our for nested operations since they're not supported by the SketchUp API. If all my code uses this wrapper, it's easy to check that the last operation ended before starting a new one, otherwise there's unwanted nesting and I emit a warning.

However, I also want to watch our for operations coming from other sources that do not use the wrapper (native operations after an action from the user, changes coming from another extension...).

That's why I ended up using the `onTransactionXXX` methods to count "third-party" operations starting (`ongoing_operations += 1`) and committing/aborting (`ongoing_operations -= 1`). Then I emit a warning if my wrapper executes a new operation while `ongoing_operations > 0`, which means that we're probably messing with the undo stack in an undesirable way.

It seems that this won't be possible due to this issue.
Is there any other way to achieve this?

Thanks

**Comment by DanRathbun** [comment]
> Is that expected behavior?

**YES**.

_From the [SketchUp Ruby API Release Notes](https://ruby.sketchup.com/file.ReleaseNotes.html#label-Observer+Upgrades) ..._
> #### Observer Upgrades
>We have been working on finding ways to make observers more stable and reliable for all developers. Most of these changes happen under the hood and are not entirely visible to developers. To help with that, we added a few Ruby Warnings that appear in the Ruby Console, and two debugging methods that will determine whether those warnings are shown: [`Sketchup.debug_mode=`](https://ruby.sketchup.com/Sketchup.html#debug_mode=-class_method) and [`Sketchup.debug_mode?`](https://ruby.sketchup.com/Sketchup.html#debug_mode%3F-class_method).
>
> * Observer events are now queued up until the active operation is done.
>
> * [`Sketchup::ModelObserver`](https://ruby.sketchup.com/Sketchup/ModelObserver.html) doesn't notify about intermediate Ruby operations (“transactions”).
>
> * Observers should no longer return zombie Ruby objects for entities. See the Observers 2016 pdf for more information.
>
> Removed dummy methods from all observer base classes in order to improve performance.
See [this pdf](https://assets.sketchup.com/files/ewh/Observers2016.pdf) for a more in depth look at the changes we made to the Observers.

 Please read the SketchUp 2016 Observer overhaul guide ...

* https://assets.sketchup.com/files/ewh/Observers2016.pdf

**Comment by DanRathbun** [comment]
> I'm writing a wrapper to make SketchUp operations a little bit safer ...

The SketchUp Team did something for versions **prior to 2016**. (Do not use for later versions.)

* [**Safer SketchUp Observer Events** (for SketchUp 2015 or older)](https://github.com/SketchUp/sketchup-safe-observer-events)

----


> ... with an API like
> 
> ```ruby
> ModelOperation.do(some options) {
>   # update the model here
> }
> ```

This is called a **block form method**, and it has been requested by multiple coders in the past (including **myself** back in 2012.)

See:

[Issue 52 : **Model#start_operation taking optional code block**](https://github.com/SketchUp/api-issue-tracker/issues/52)
... in which I posted an example:
```ruby
    def with_operation(op_name, disable_ui = true)
      return unless block_given?
      model = Sketchup.active_model
      model.start_operation(op_name, disable_ui)
        #
        result = yield # call the block
        #
      model.commit_operation
    rescue => err
      puts err.inspect
      model.abort_operation if UI.messagebox(
        "There was an error:\n#{err.message}\n\nDo you wish to Undo ?", MB_YESNO
      ) == IDYES
    else
      return result
    end
```

----

There was also discussion in the public forum ...

* [Open operations](https://forums.sketchup.com/t/open-operations/154632)

... and posted a different example there.

**Comment by merwaaan** [comment]
@DanRathbun Thanks for the thorough reply.
Glad to see that other developers have already discussed such ideas at length.

Because developers who are not familiar with the history of SketchUp (myself included) are unlikely to stumble upon these documents, the documentation of those transaction event handlers should be updated to incorporate this information.

**Comment by DanRathbun** [comment]
> Because developers who are not familiar with the history of SketchUp (myself included) are unlikely to stumble upon these documents, the documentation of those transaction event handlers should be updated to incorporate this information.

I would agree that at the least the blurb from the release notes (which contains the link to the PDF file,) should be added to the class overview and perhaps a new "Observers" section on the "SketchUp Ruby API" [index page](https://ruby.sketchup.com/index.html).

It is hidden too far down within the release notes for new coders to find.

**Comment by thomthom** [comment]
I'm logging this as a documentation issue to clarify how/when observers are triggered and making it easier to find that doc.

**Comment by thomthom** [comment]
> I also want this wrapper to watch our for nested operations since they're not supported by the SketchUp API. If all my code uses this wrapper, it's easy to check that the last operation ended before starting a new one, otherwise there's unwanted nesting and I emit a warning.
> 
> However, I also want to watch our for operations coming from other sources that do not use the wrapper (native operations after an action from the user, changes coming from another extension...).

What would you do if you were able to detect something else using `model.start_operation`?

**Comment by merwaaan** [comment]
@thomthom In that case, I would emit a warning too, so that we're at least aware that there's some unexpected nesting going on, possibly wreaking havok on the model.

Thanks to the wrapper described in the first post, we can detect conflicts between two operations triggered by our extension.

If we were able to also monitor "external" operations, then all the bases would be covered and we would be extra safe.

**Comment by DanRathbun** [comment]
One of the things that has been requested in the past was a boolean [`model.operation_open?`]() method.

Also I think Ken has politicked for a [`model.current_operation_name`]() method.

**Comment by merwaaan** [comment]
@DanRathbun  Some way to query the model about any ongoing operation would be great indeed.
